### PR TITLE
4115 Combining 'indexMissingFields' and 'advancedHSearchIndexing' functionalities will prevent creating a resource

### PR DESCRIPTION
**Pull request by epeartree**

### CODE REVIEW

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java [hunk 1]
```diff
@@ -59,6 +64,33 @@ void testExtract_composite_producesValues() {
 		assertThat(spIndexData, hasSize(1));
 	}
 
+	//TODO EP: Write a test which enables both settings, and then: Creates a patient with gender male, creates a patient with no gen0.der. And then searches on patient?gender:missing=true
+	@Test
+	void testExtractWithMissingParams() {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java:69]: this is a good start for a name.  let's see if we can make it more descriptive:
we have agreed upon a format where we specify what is being tested, under which condition and the expected result.

testExtract_withParamMarkedAsMissing_willBeIgnored

in this format, know what being tested, and even more important, what should be the result of the test, without any doubt.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java [hunk 2]
```diff
@@ -59,6 +64,33 @@ void testExtract_composite_producesValues() {
...
+		ResourceIndexedSearchParamDate searchParamDate = new ResourceIndexedSearchParamDate(new PartitionSettings(), "SearchParameter", "Date", null, null, null, null, null);
+		searchParamDate.setMissing(true);
+		searchParams.myDateParams.add(searchParamDate);
+
+		ResourceIndexedSearchParamQuantity searchParamQuantity = new ResourceIndexedSearchParamQuantity(new PartitionSettings(), "SearchParameter","Quantity", null, null,null);
+		searchParamQuantity.setMissing(true);
+		searchParams.myQuantityParams.add(searchParamQuantity);
+
+		// run: now translate to HSearch
+		ResourceSearchParams activeSearchParams = mySearchParamRegistry.getActiveSearchParams("Observation");
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java:81]: I would use a 'Patient' to make the test more relevant to the issue we're trying to solve.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java [hunk 3]
```diff
@@ -59,6 +64,33 @@ void testExtract_composite_producesValues() {
...
+		ExtendedHSearchIndexExtractor extractor = new ExtendedHSearchIndexExtractor(
+			myDaoConfig, myFhirContext, activeSearchParams, mySearchParamExtractor, myModelConfig);
+		ExtendedHSearchIndexData indexData = extractor.extract(new SearchParameter(), searchParams);
+
+		// validate
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java:86]: I like the comments separating the phases.  it helps the reader to eyes to get on what matters.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java [hunk 4]
```diff
@@ -59,6 +64,33 @@ void testExtract_composite_producesValues() {
 		assertThat(spIndexData, hasSize(1));
 	}
 
+	//TODO EP: Write a test which enables both settings, and then: Creates a patient with gender male, creates a patient with no gen0.der. And then searches on patient?gender:missing=true
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractorTest.java:67]: now that we have a unit test, we kind of need an integrated test to test the above.  and when we'll have that test, we can remove the above.

**Review by epeartree — COMMENTED** [review]
Good test!

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 5]
```diff
@@ -6095,6 +6095,30 @@ public void testTransaction() throws Exception {
 		}
 	}
 
+	@Test
+	public void testTransactionWithMissingParamsAndAdvancedHSearchEnabled() throws Exception {
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:6099]: testCreatePatientWithAdvancedHSearchIndexingAndIndexMissingFieldsEnableSucceeds?

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 6]
```diff
@@ -6095,6 +6095,30 @@ public void testTransaction() throws Exception {
...
+		myDaoConfig.setAdvancedHSearchIndexing(true);
+		myDaoConfig.setIndexMissingFields(DaoConfig.IndexEnabledEnum.ENABLED);
+		String methodName = "testTransactionWithMissingParamsAndAdvancedHSearchEnabled";
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:6102]: i'd call this variable identifierValue and assign in "someValue"

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 7]
```diff
@@ -6095,6 +6095,30 @@ public void testTransaction() throws Exception {
...
+		Patient patient = new Patient();
+		patient.addIdentifier().setSystem("urn:system").setValue(methodName);
+		patient.addName().setFamily(methodName).addGiven(methodName);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:6105]: no need to do a addName.  let's keep it to the mininum of what we need.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 8]
```diff
@@ -6095,6 +6095,30 @@ public void testTransaction() throws Exception {
...
+
+		String resource = myFhirContext.newXmlParser().encodeResourceToString(patient);
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:6107]: line 6107 all the way to 6119 is difficult and cumbersome.  we have our super handy 'myClient' that does a wonderful job.  see test below for the perfect example.
 except that u don't need to wrap the test in a try catch.  all u would need to do is check for return value of 201(http).  there's examples in this class somewhere.
 
using myClient will set us up nicely for doing a search with a modifier of ':missing'.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java [hunk 9]
```diff
@@ -6095,6 +6095,30 @@ public void testTransaction() throws Exception {
...
+		IdType id;
+		try {
+			assertEquals(201, response.getStatusLine().getStatusCode());
+			String newIdString = response.getFirstHeader(Constants.HEADER_LOCATION_LC).getValue();
+			assertThat(newIdString, startsWith(ourServerBase + "/Patient/"));
+		} finally {
+			response.close();
+		}
+	}
+
```
Comment (epeartree) [inline hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/provider/r4/ResourceProviderR4Test.java:6121]: Good first kick at the can.  i suggest finding out the type of myClient, where it's defined and then, go and look it up in the Hapi-fhir documentation.  or even better, go read about it before u do the mods suggested above.

When u'r all good, ping me so we can talk about implenting the solution.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4115-combining-indexmissingfields-and-advancedhsearchindexing-functionalities-will-prevent-creating-a-resource.yaml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+---
+type: fix
+issue: 4115
+jira: SMILE-3559
```
Comment (epeartree) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4115-combining-indexmissingfields-and-advancedhsearchindexing-functionalities-will-prevent-creating-a-resource.yaml:4]: There is no JIRA ticket associated with this issue.  please remove and DM me about.

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4115-combining-indexmissingfields-and-advancedhsearchindexing-functionalities-will-prevent-creating-a-resource.yaml [hunk 11]
```diff
@@ -0,0 +1,8 @@
...
+title: "Previously, enabling 'indexMissingFields' and 'advancedHSearchIndexing'
```
Comment (epeartree) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4115-combining-indexmissingfields-and-advancedhsearchindexing-functionalities-will-prevent-creating-a-resource.yaml:5]: Excellent description.

File: hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchIndexData.java [hunk 12]
```diff
@@ -141,6 +141,7 @@ public void addDateIndexData(String theSpName, DateSearchIndexData value) {
 		mySearchParamDates.put(theSpName, value);
 	}
 
+	public SetMultimap<String, DateSearchIndexData> getDateIndexData() { return mySearchParamDates; }
```
Comment (epeartree) [inline hapi-fhir-jpaserver-model/src/main/java/ca/uhn/fhir/jpa/model/search/ExtendedHSearchIndexData.java:144]: please add a space in between method

**Review by epeartree — COMMENTED** [review]
Well done!

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4115-combining-indexmissingfields-and-advancedhsearchindexing-functionalities-will-prevent-creating-a-resource.yaml (hunk 11 above)
Comment (karneet1212) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_2_0/4115-combining-indexmissingfields-and-advancedhsearchindexing-functionalities-will-prevent-creating-a-resource.yaml:4]: Thank you

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractor.java [hunk 13]
```diff
@@ -96,9 +96,13 @@ public ExtendedHSearchIndexData extract(IBaseResource theResource, ResourceIndex
 		theNewParams.myNumberParams.forEach(nextParam ->
 			retVal.addNumberIndexDataIfNotPresent(nextParam.getParamName(), nextParam.getValue()));
 
-		theNewParams.myDateParams.forEach(nextParam -> retVal.addDateIndexData(nextParam.getParamName(), convertDate(nextParam)));
+		theNewParams.myDateParams.stream()
```
[the remark this answers is not in the document]
Comment (karneet1212) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/search/ExtendedHSearchIndexExtractor.java:102]: The reason was that only date and quantity were being affected and were causing issues out of all of the parameters. However to keep it consistent and allow all of the missing parameters to be dealt with together, I have now added the change to all of the other parameters.

**Review by jamesagnew — APPROVED** [review]

**Pull request merged**